code/code/misc/offense.cc

```diff
@@ -766,332 +766,301 @@ static bool canFleeThisWay(TBeing *ch, dirTypeT dir)
+  */
+  spellNumT retreatSkill = getSkillNum(SKILL_RETREAT);
+  bool doesKnowRetreat = doesKnowSkill(retreatSkill);
+  bool wasRetreatSuccessful = doesKnowRetreat && bSuccess(retreatSkill);
+
+  bool panic = false;
+  int rc = 0;
+  auto *riderAsTBeing = dynamic_cast<TBeing *>(rider);
+
   if (riding) {
-    // Let's make retreat skill have some chance of keeping you on the mount - Brutius 07/26/1999
-    if (!(doesKnowSkill(skill) && doesKnowSkill(skill2)) || 
-	!(bSuccess(skill) && bSuccess(skill2))) { 
-      sendTo("Your panic causes you to fall.\n\r");
+    // Allow successful retreat or ride checks to avoid falling off mount. Deikhans
+    // get extra attempt to remain mounted via chivalry.
+    if ((doesKnowSkill(SKILL_RIDE) && !bSuccess(SKILL_RIDE)) ||
+        (doesKnowSkill(SKILL_CHIVALRY) && !bSuccess(SKILL_CHIVALRY)) || !wasRetreatSuccessful) {
+      act("You urge your $O to flee, causing it to panic and throw you off!", true, this,
+          nullptr, riding, TO_CHAR);
+      act("$n urges $s mount to flee, causing it to panic and throw $n off!", true, this,
+          nullptr, riding, TO_ROOM);
       rc = fallOffMount(riding, POSITION_SITTING);
-      panic = TRUE;
-      if (IS_SET_DELETE(rc, DELETE_THIS)) 
+      if (IS_SET_DELETE(rc, DELETE_THIS))
         return DELETE_THIS;
+      panic = true;
     }
   } else if (getPosition() <= POSITION_SITTING) {
     addToMove(-10);
-    act("$n scrambles madly to $s feet!", TRUE, this, 0, 0, TO_ROOM);
-    act("Panic-stricken, you scramble to your feet.", TRUE, this, 0, 0, TO_CHAR);
+    // If retreat succeeded, simply being on the ground when fleeing doesn't
+    // result in panic. And you get a different message.
+    if (wasRetreatSuccessful) {
+      act("$n rolls aside and regains $s footing!", true, this, nullptr, nullptr, TO_ROOM);
+      act("You quickly roll aside and regain your footing.", true, this, nullptr, nullptr, TO_CHAR);
+    } else {
+      act("$n scrambles madly to $s feet!", true, this, nullptr, nullptr, TO_ROOM);
+      act("Panic-stricken, you scramble to your feet.", true, this, nullptr, nullptr, TO_CHAR);
+    }
+
     doStand();
-    addToWait(combatRound(1));
 
-    panic = TRUE;
-  } else if ((t = rider)) {
-    t->sendTo("Your mount panics and attempts to flee.\n\r");
-    rc = t->fallOffMount(this, POSITION_SITTING);
-    panic = TRUE;
+    // If retreat didn't succeed or char doesn't have it, add some lag
+    // and panic.
+    if (!wasRetreatSuccessful) {
+      addToWait(combatRound(1));
+      panic = true;
+    }
+    // If <this> is a mount being ridden
+  } else if (riderAsTBeing) {
+    act("Your $O panics and attempts to flee!", true, riderAsTBeing, nullptr, this, TO_CHAR);
+    act("$n's $O panics and attempts to flee!", true, riderAsTBeing, nullptr, this, TO_ROOM);
+
+    // Allow deikhans to pass checks and prevent getting bucked
+    if (riderAsTBeing->doesKnowSkill(SKILL_CHIVALRY) && riderAsTBeing->bSuccess(SKILL_CHIVALRY) &&
```

Comment:
why doesn't bSuccess automatically check for skill knowledge?

Reply:
It does, actually, and I'm not sure why we don't rely on it more in the code. Addressed this in https://github.com/sneezymud/sneezymud/pull/373/commits/9eafa770ec48ccd42c3287b72176a0950acf293e.